pkg/controllers/plugin/pluginpreset_controller.go

```diff
@@ -0,0 +1,175 @@
+		}
+		return ctrl.Result{}, nil
+
+	}
+	if err := clientutil.EnsureFinalizer(ctx, r.Client, pluginPreset, greenhouseapis.FinalizerCleanupPluginPreset); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	clusters, err := r.listClusters(ctx, pluginPreset)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// TODO: GC plugins that are not in the list of clusters anymore
+	plugins, err := r.listPlugins(ctx, pluginPreset)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := r.cleanupPlugins(ctx, clusters, plugins); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := r.reconcilePluginPreset(ctx, pluginPreset, clusters); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// cleanupPlugins deletes all plugins where the clusterName is not included in the ClusterList.
+func (r *PluginPresetReconciler) cleanupPlugins(ctx context.Context, clusters *greenhousev1alpha1.ClusterList, plugins *greenhousev1alpha1.PluginList) error {
+	var allErrs = make([]error, 0)
+	for _, plugin := range plugins.Items {
+		if !slices.ContainsFunc(clusters.Items, func(cluster greenhousev1alpha1.Cluster) bool {
+			return cluster.GetName() == plugin.Spec.ClusterName
+		}) {
+			if err := r.Client.Delete(ctx, &plugin); err != nil && !apierrors.IsNotFound(err) {
+				allErrs = append(allErrs, err)
+			}
+		}
+	}
+	return errors.Join(allErrs...)
+}
+
+func (r *PluginPresetReconciler) reconcilePluginPreset(ctx context.Context, pb *greenhousev1alpha1.PluginPreset, clusters *greenhousev1alpha1.ClusterList) error {
+	var allErrs = make([]error, 0)
+
+	for _, cluster := range clusters.Items {
+		plugin := &greenhousev1alpha1.Plugin{
+			ObjectMeta: v1.ObjectMeta{
+				Name:      generatePluginName(pb.Spec.Plugin.PluginDefinition, &cluster),
+				Namespace: pb.GetNamespace(),
+			},
+		}
+		result, err := clientutil.CreateOrPatch(ctx, r.Client, plugin, func() error {
+			// Label the plugin with the managed resource label to identify it as managed by the PluginPreset.
+			plugin.SetLabels(map[string]string{greenhouseapis.LabelKeyPluginPreset: pb.Name})
+			// Set the owner reference to the PluginPreset. This is used to trigger reconciliation, if the managed Plugin is modified.
+			if err := controllerutil.SetControllerReference(pb, plugin, r.Scheme()); err != nil {
+				return err
+			}
+			plugin.Spec = pb.Spec.Plugin
+			// Set the cluster name to the name of the cluster. The PluginSpec contained in the PluginPreset does not have a cluster name.
+			plugin.Spec.ClusterName = cluster.GetName()
+			return nil
+		})
+		switch result {
+		// TODO: Handle the result. Log and emit event.
+		}
+		allErrs = append(allErrs, err)
+	}
+	return utilerrors.NewAggregate(allErrs)
+}
+
+// generatePluginName generates a name for a plugin based on the used PluginDefinition and the Cluster.
+func generatePluginName(pluginDefinition string, cluster *greenhousev1alpha1.Cluster) string {
+	return fmt.Sprintf("%s-%s", pluginDefinition, cluster.GetName())
```

Comment:
Only a remark on the naming here on whether to use `<pluginPreset.name>-<cluster.name>` to make it clearer what plugins belong to it.
Though, I see that `<pluginDefinition.name>-<cluster.name>` is also good to figure out what the actual plugin does.

What would happen, if a user already created a plugin with any of the naming conventions before creating the plugin preset?

Reply:
> Only a remark on the naming here on whether to use <pluginPreset.name>-<cluster.name> to make it clearer what plugins belong to it.
Though, I see that <pluginDefinition.name>-<cluster.name> is also good to figure out what the actual plugin does.

I like it. That makes it more obvious in the Plugin UI. The `kubectl get plugins` already contains both the _PluginDefinition_ and the _Cluster_ in the result.  

> What would happen, if a user already created a plugin with any of the naming conventions before creating the plugin preset?

Good point. I had not considered this yet. As a safeguard we could skip existing Plugins that do not have the `OwnerReference` and/or the `greenhouse.sap/pluginpreset` set.